File: backend/app/schemas/calculator.py
from pydantic import BaseModel
from enum import Enum
from app.schemas.insurance import InsurancePlan


class ServiceType(str, Enum):
    office_visit = "office_visit"
    specialist = "specialist"
    urgent_care = "urgent_care"
    er = "er"
    prescription = "prescription"
    procedure = "procedure"
    preventive = "preventive"
    lab = "lab"
    imaging = "imaging"


class CostScenario(BaseModel):
    service_type: ServiceType
    service_name: str
    retail_price: float
    frequency: int


class CostCalculationRequest(BaseModel):
    insurance_plan: InsurancePlan
    scenarios: list[CostScenario]


class CostLineItem(BaseModel):
    service_name: str
    retail_price: float
    frequency: int
    annual_retail_cost: float
    annual_insured_cost: float
    deductible_applied: float
    copay_applied: float
    coinsurance_applied: float
    insurance_paid: float
    savings: float


class CostCalculationResult(BaseModel):
    line_items: list[CostLineItem]
    totals: dict
    recommendation: str

File: frontend/src/types/calculator.ts
import { InsurancePlan } from "./insurance";

export type ServiceType = "office_visit" | "specialist" | "urgent_care" | "er" | "prescription" | "procedure" | "preventive" | "lab" | "imaging";

export interface CostScenario {
  service_type: ServiceType;
  service_name: string;
  retail_price: number;
  frequency: number;
}

export interface CostCalculationRequest {
  insurance_plan: InsurancePlan;
  scenarios: CostScenario[];
}

export interface CostLineItem {
  service_name: string;
  retail_price: number;
  frequency: number;
  annual_retail_cost: number;
  annual_insured_cost: number;
  deductible_applied: number;
  copay_applied: number;
  coinsurance_applied: number;
  insurance_paid: number;
  savings: number;
}

export interface CostCalculationResult {
  line_items: CostLineItem[];
  totals: {
    annual_premiums: number;
    annual_retail_total: number;
    annual_insured_total: number;
    annual_insurance_pays: number;
    deductible_used: number;
    deductible_remaining: number;
    oop_max_remaining: number;
    total_true_cost_insured: number;
    net_savings: number;
    insurance_value_ratio: number;
  };
  recommendation: string;
}
